Skipping empty write requests

schmidt-sebastian opened the pull request:
This is required for a full pass of the conformance test.

This adds fromObject and fromUpdateMap to DocumentTransform, DocumentMask, and DocumentSnapshot and moves some of the existing logic to expand objects into these functions.

This also adds the ability to delete nested fields via merge. set({a : { b:  delete}}, merge) was previously not supported, but I needed it for the system test.

mikelehen commented:
Out of curiosity, could we use a JSON representation of the protos or something for sane diagnosability of failures, etc.?

schmidt-sebastian commented on conformance/runner.js:
```diff
@@ -161,8 +158,12 @@ function runTest(spec) {
-      resolve();
+      try {
+        assert.deepEqual(request, expected);
+        resolve();
```
I reworked this to take advantage of the error handling that is built into the SDK. This means that I now have to return proper responses from the request handlers, but I moved all this logic to a single place.

I added the try/catch originally to handle unhandled promise rejections. The asserts already failed the tests, but returned unhandled Promised back to Node. Since everything is wrapped in Promise.resolve() and request blocks now, this special handling is no longer necessary.

schmidt-sebastian commented on src/document.js:
```diff
@@ -191,6 +191,95 @@ class DocumentSnapshot {
+   * turns { foo.bar : foobar } into { foo { bar : foobar }}
+   *
+   * @private
+   * @param {firestore/DocumentReference} ref - The reference to the documen.t
```
Done

schmidt-sebastian commented on src/document.js:
```diff
@@ -480,6 +569,19 @@ class DocumentSnapshot {
+      return true;
+    }
+
+    return isEmptyObject(this._fieldsProto);
```
Sounds good.

schmidt-sebastian commented on src/write-batch.js:
```diff
@@ -463,22 +459,41 @@ class WriteBatch {
     }
 
+    // We store the number of operations that we sent as part of this commit
+    // as the user is able to add additional operations afterwards.
```
I added this, but this is yet another breaking API change (which hopefully won't actually break anyone). Before this PR, calling commit() twice didn't really need any special handling.

I also decided to let users call commit() multiple times (as long as they don't modify the batch. This makes retrying after request failures much easier.

schmidt-sebastian commented:
Addressed comments. I agree that this PR brings very little benefit to Node, but I think the benefit that it brings to the GCloud SDKs as a whole far outweighs the work that has gone into this.

mikelehen commented on src/write-batch.js:
```diff
@@ -490,9 +510,7 @@ class WriteBatch {
-    // We store the number of operations that we sent as part of this commit
-    // as the user is able to add additional operations afterwards.
-    let opCount = this._writes.length;
+    this._committed = true;
```
It might be worth adding a comment (to commit() or the top of commit_() calling out that we're explicitly not calling verifyNotCommitted() because we want to let people retry committing batches.  Else somebody might come along and assume it's an oversight and "fix" it. :-)

schmidt-sebastian replied:
Done.

mikelehen commented on test/write-batch.js:
```diff
@@ -294,6 +294,21 @@ describe('batch support', function() {
       });
   });
 
+  it('cannot append to committed batch', function() {
```
Can you add a test that verifies that you can call commit() multiple times if that is the contract we're offering?

schmidt-sebastian replied:
Done.
